### ISSUE TITLE
[FEATURE] Folder Structure

### ISSUE THREAD

**Issue by Silther** [stumpapp/stump#254]
**Is your feature request related to a problem? Please describe.**
My Folder structure is something like:
```
Library 1
|-- Author 1
|   |-- Series 1
|   |   |-- Book 1
|   |   |   |-- Book 1.epub
|   |   |   |-- Cover.jpg
|   |   |   |-- Metadata
|   |   |-- Book 2
|   |   |   |-- Book 2.epub
|   |   |   |-- Cover.jpg
|   |   |   |-- Metadata
|   |-- Series 2
|   |   |-- ...
|-- Author 2
|   |-- ...
Library 2
|-- ...
```
The folder structure is used by Audiobookshelf and is pretty similar to Calibre.
The problem is, that Stump reads the Series only from the the folder Structure and not the metadata, which leads either to every book being in  it's on series, or the author being the series, both is not ideal.

**Describe the solution you'd like**
Add the folder structure as a third option, or only read series from the metadata

**Describe alternatives you've considered**
probably the easiest option would be to allow books be placed in a book-folder and the book-folder into a series-folder

**Comment by aaronleopold** [comment]
> The problem is, that Stump reads the Series only from the the folder Structure and not the metadata, which leads either to every book being in it's on series, or the author being the series, both is not ideal.
> ...
> Add the folder structure as a third option, or only read series from the metadata

You're correct, Stump does not currently support building a library from metadata. Rather, it is the file structure. The metadata is used afterwards, e.g. filtering. I am not necessarily opposed to adding support for the former (library from metadata), however the requirements for how it should work would need to be thoroughly laid out (namely, what does the metadata look, where is valid placement throughout the _different_ file structures, and how should Stump reconcile parsed metadata with what it indexes on disk).

As for supporting the other file structure explicitly, I'll consider it. To be completely honest, assuming I am understanding correctly, it is almost exactly what the `Collection priority` pattern achieves. The only difference is that, in your case, you'd like the first child directory of the library root to be ignored. Supporting that in and of itself is not too much, but it just means supporting yet another pattern to maintain. I am currently rewriting the scanner, so I'll have a better idea towards the end of that effort what supporting that structure might look like.

As a side note, I think I'll need catchier names for these patterns because I can't think of what this proposed one would be called relative to what Stump supports today (`Collection` vs `Series` priority) 😅

**Comment by aaronleopold** [comment]
> A recursive library there every book is inside its own folder more would make more sense to me as it seems more future proof

This is a prerequisite for the linked ticket (#196). The idea for that one is you can group multiple files of the same book within a single folder, e.g. a PDF and an EPUB, and Stump would recognize the different files while retaining a single media entry in the DB to group them. If it can do that, then it can recognized a folder containing parts of a book. The long word vomit I had above was just a tangent off of that idea that I thought was similar-ish.

> on a related note, why is one library restricted to one folder and one folder restricted to one library?

I'm not sure what you mean. Libraries don't currently support multiple sources, if that is the question. Hypothetically it can one day be extended to support it, I imagine it like how Plex works. However, for the second part of your question, why would you want the same folder present for multiple libraries? You would have duplicated media

**Comment by cathedralnarwhal** [comment]
Just wanted to say that I'm interested in support for nested series. Thank you for your hard work.

**Comment by tenninjas** [comment]
I think nested series are the best way to go.
I also just want to add a +1 as a user who would definitely use multiple sources for a single library.

**Comment by aaronleopold** [comment]
> I think nested series are the best way to go. I also just want to add a +1 as a user who would definitely use multiple sources for a single library.

I created a [separate issue](https://github.com/stumpapp/stump/issues/280) last week to better track the request 🙂 I like the idea so am planning on tackling it hopefully before `0.2.0`. I'll be focusing on general stability and honing the existing features for `0.1.0`

I also want to say that I've been working on some usability improvements for the file explorer, extending it to be re-used for both library and series screens. Since it is just traversing the file system at the relative root, I think it close to what nested series might look like until it gets worked. So it could be a valid alternative until that point. Currently, those changes are in `experimental`, but will make its way to `nightly` and  `0.0.2` shortly after.

**Comment by beville** [comment]
I wanted to throw in a +1 request for supporting imports where the series is based only on comic file metadata (internal tags firstly, then filename parsing), rather than any folder location.  This would useful for folks who have meticulously tagged files that aren't organized into a particular folder structure.  

Currently  I think Codex is the only server app that supports this, but it would be nice to have more options.

(As a experiment, though, I have a test script to build a folder tree based on publisher/series tags, and populate with symlinks to the actual file locations.  It's functional with Stump, but less than ideal.  )

**Comment by aaronleopold** [comment]
I've created a separate issue for folks to better track it: https://github.com/stumpapp/stump/issues/288

**Comment by beville** [comment]
Thanks for considering this!

Re: Kavita, I just had another look to be sure, and seems that it still relies on a strict folder structure: https://wiki.kavitareader.com/en/guides/managing-your-files/scanner#introduction

**Comment by Critical-Impact** [comment]
Komga does read metadata as well, https://komga.org/docs/guides/scan-analysis-refresh/
Has a brief rundown of what fields it expects from the metadata files it can read

**Comment by aaronleopold** [comment]
> Komga does read metadata as well, https://komga.org/docs/guides/scan-analysis-refresh/
> 
> Has a brief rundown of what fields it expects from the metadata files it can read

Komga does read metadata, but I don't think there exists an option to completely ignore the underlying filesystem and instead rely solely on metadata to construct the library and series (which is the ask discussed above)

**Comment by ecraven** [comment]
I have about 50000 PDFs and 20000 epub/mobis, I'd love to use Stump with them, but a single-level hierarchy won't work for this :-/ I'd love to have a nested hierarchy, based on the directory structure (with additional metadata read from the file and if possible at all also the filename).

**Comment by kulak** [comment]
> I have about 50000 PDFs and 20000 epub/mobis, I'd love to use Stump with them, but a single-level hierarchy won't work for this :-/ I'd love to have a nested hierarchy, based on the directory structure (with additional metadata read from the file and if possible at all also the filename).

I made the same proposal here:

https://github.com/stumpapp/stump/discussions/522

I tried to put a book and found out that it does not fit a single category structure.  

Tagging cannot belong to Library.  I initially was going to have a library per language, but then I realized that major feature of library is tagging.  So, I had to break a single library into a set of smaller libraries and to impose a primary categorization based on the file system.

Editing epub files is not an option.   The only option I see is having metadata TOML file per file and per series.

**Comment by ecraven** [comment]
Well, ideally I'd love to have *only* tags, I don't need a hierarchy per se.

I am strictly opposed to putting any data *away* from the file it belongs to, so metadata inside the file or the filename are the only options I am willing to entertain for myself, because otherwise it *always* gets separated or lost. If it is part of the actual file, that can't happen easily, and I won't get locked into one application.

I have no say in what Stump does, but I'd much prefer at least an *option* to read this data from the file directly. I am willing to tag all my pdfs and epubs (not necessarily via Stump, I like that it can work with read-only access, though I wouldn't be averse to using it to add tags, but I understand that *writing* data is always very dangerous).

Thanks for discussing this!

**Comment by aaronleopold** [comment]
> I'd love to have a nested hierarchy, based on the directory structure

@ecraven A nested hierarchy based on the directory structure is effectively https://github.com/stumpapp/stump/issues/280. Unless you mean something else?

@kulak I responded directly in your discussion 👍 

It seems that the discussion here is moving away from folder structure and more towards tags, which is not necessarily on topic for this issue. I'll respond to a few bits, but let's keep any future tag-management-specific topics to tag-management-specific issues if they exist (or create one if you feel an existing one doesn't cover your feature).

> If it is part of the actual file, that can't happen easily, and I won't get locked into one application

This is also my personal preference, I wrote something similar in the linked discussion post

> I'd much prefer at least an option to read this data from the file directly

Stump already reads metadata from files, if present. See https://www.stumpapp.dev/guides/basics/books#metadata. I think a feature request that could be made is to better support tags from these metadata files.

**Comment by ecraven** [comment]
Sorry about mixing tags in here, I'll keep that discussion to #151. 

I'm not sure whether I understand #280 correctly. In my case, it happens that I have the *same* subdirectory in multiple parts of the hierarchy, so just creating collections based on the direct parent directory wouldn't work, the *entire* path to the directory should be the name of the series.

So a file `/computers/compilers/optimizations/TCE.pdf` should not be in series `optimizations`, but in series `computers - compilers - optimizations`. Is that what that issue is suggesting?

I'll look into what options there are for tags in PDF, and come back with a concrete suggestion for what to read from the file, does that sound reasonable?

**Comment by beville** [comment]
>     4. **Metadata Tags [Auto-tagging from nested directory tree structure #522](https://github.com/stumpapp/stump/discussions/522):** Ignore the folder structure completely and only use a metadata file stored alongside the book file (would presumably need a folder for each book).
> 

@Silther Regarding point 4 above,  my original add-on request (from March 2024) was for honoring metadata embedded into the cbz and cbr files.  For that scenario there could be 100 unrelated comics in one folder, but since each comic file could (should) contain the all it's own metadata, any folder structure would be irrelevant.

Apologies if I'm missing context, but want to make sure that idea wasn't lost.

**Comment by vmpyr** [comment]
Have we thought of reading the metadata from `metadata.opf` files?
For example, my folder structure is like this:
```
ROOT
|-- Author 1
|   |-- Book 1
|   |   |-- Book 1 - Author 1.epub
|   |   |-- cover.jpg (mentioned in opf)
|   |   |-- metadata.opf
|   |-- Book 2
|   |   |-- Book 2 - Author 1.epub
|   |   |-- cover.jpg (mentioned in opf)
|   |   |-- metadata.opf
|-- Author 2
...
```

All books exist separately. The metadata file has calibre tags about the Series Name and Book No. in the Series. Kavita works very well this (Kavita works very well with almost any structure tbh). On the Kavita dashboard, Series are listed and opening up gives all the books in the series in Series No. order from metadata. The cover image is also the jpg referenced in the `.opf` file.

I think using `.opf` metadata files should also be incorporated. Thoughts?

**Comment by aaronleopold** [comment]
@vmpyr This was actually added as part of https://github.com/stumpapp/stump/pull/607. That will be included with the next versioned release (`0.0.11`). It is also available in `nightly` or `experimental` if you wanted to try it out early.

I've been a bit slower to release with the big migration going on and then personal things (I'm in the process of moving) so apologies in general for anyone who has been waiting on things

**Comment by vmpyr** [comment]
@aaronleopold I am on the `nightly` build and the cover images and rest of the metadata works fine yes, but the series part doesn't unfortunately. We'll need to parse the `calibre:<tag>` metadata tags for this. Kavita parses those tags already since calibre style of metadata is almost a standard for a lot of digital libraries now.
An example:
```xml
<?xml version='1.0' encoding='utf-8'?>
<package xmlns="http://www.idpf.org/2007/opf" unique-identifier="uuid_id" version="2.0">
    <metadata xmlns:dc="http://purl.org/dc/elements/1.1/" xmlns:opf="http://www.idpf.org/2007/opf">
        <dc:identifier opf:scheme="calibre" id="calibre_id">106</dc:identifier>
        <dc:identifier opf:scheme="uuid" id="uuid_id">373c64ba-39fd-40b3-99ba-0223ebab0fec</dc:identifier>
        <dc:title>After the Funeral</dc:title>
        <dc:creator opf:file-as="Christie, Agatha" opf:role="aut">Agatha Christie</dc:creator>
        <dc:contributor opf:file-as="calibre" opf:role="bkp">calibre (8.2.1) [https://calibre-ebook.com]</dc:contributor>
        <dc:date>1953-02-28T18:30:00+00:00</dc:date>
        <dc:description>&lt;div&gt;
&lt;p&gt;The master of a Victorian mansion dies suddenly – and his sister is convinced it was murder…. When Cora is savagely murdered with a hatchet, the extraordinary remark she made the previous day at her brother Richard’s funeral suddenly takes on a chilling significance. At the reading of Richard’s will, Cora was clearly heard to say: ‘It’s been hushed up very nicely, hasn’t it…But he was murdered, wasn’t he?’ In desperation, the family solicitor turns to Hercule Poirot to unravel the mystery.&lt;/p&gt;&lt;/div&gt;</dc:description>
        <dc:publisher>HarperCollins</dc:publisher>
        <dc:identifier opf:scheme="ISBN">9780007562695</dc:identifier>
        <dc:identifier opf:scheme="AMAZON">0007562691</dc:identifier>
        <dc:identifier opf:scheme="GOODREADS">60458674</dc:identifier>
        <dc:language>eng</dc:language>
        <dc:subject>Mystery</dc:subject>
        <dc:subject>Crime</dc:subject>
        <dc:subject>Classics</dc:subject>
        <dc:subject>Thriller</dc:subject>
        <dc:subject>Detective</dc:subject>
        <meta name="calibre:series" content="Hercule Poirot"/>
        <meta name="calibre:series_index" content="33"/>
        <meta name="calibre:rating" content="8"/>
        <meta name="calibre:timestamp" content="2025-04-14T15:01:21.815534+00:00"/>
        <meta name="calibre:title_sort" content="After the Funeral"/>
    </metadata>
    <guide>
        <reference type="cover" title="Cover" href="cover.jpg"/>
    </guide>
</package>
```

>I've been a bit slower to release with the big migration going on and then personal things (I'm in the process of moving) so apologies in general for anyone who has been waiting on things

Take your own sweet time dude, you are doing great work. I will surely help on this project once I sort some personal life goals I'm working on for a few months.

**Comment by aaronleopold** [comment]
> We'll need to parse the calibre:<tag> metadata tags for this

Ahhh okay, I understand. Thankfully that shouldn't actually too much work. Would you mind capturing this in a separate issue? Your example is also perfect

And also thank you, I appreciate it 🙂

**Comment by yann117** [comment]
Hello - I was looking through the issues to find out if mine was already reported, and here I can see:

> @Silther Regarding point 4 above, my original add-on request (from March 2024) was for honoring metadata embedded into the cbz and cbr files

My issue is precisely that I observed Stump is ignoring the embedded, fully standardised, _ComicInfo.xml_ files from my Comics, if they are named _.cbz_, which is also the common standard for Comics. On the contrary, if I rename them to _.zip_ then, it properly reads it, but then Kavita will do the opposite and ignore the _ComicInfo.xml_.

So at this point, I cannot get both software to work together on the same Comics library, which is unfortunate since both have advantages/disadvantages.

I don't see the point of ignoring on purpose the ComicInfo.xml, as it was created precisely for defining a standard in comics definition, that makes a library completely independent from the storage structure. One could decide any naming convention for the folders and comics files, which make it quite adjustable for any requirements.

WDYT?

**Comment by aaronleopold** [comment]
@yann117 Sorry this got lost in my notifications. This sounds more like a bug, Stump should read `ComicInfo.xml` files from any supported archive format

**Comment by yann117** [comment]
@aaronleopold Thanks for the feedback.
I figured it out, in fact it was two mistakes on my side:
1. I thought Stump would also display the Writer/Penciller while displaying a comic details in the info page, but it is not implemented yet
2. I used to put some text into the `<Volume>` attribute, in order to fix an issue with Kavita that otherwise create me one volume per comic into a Series, and in Stump this field is restricted to _Number_, so it was failing processing the _ComicInfo.xml_

**Comment by aaronleopold** [comment]
> I thought Stump would also display the Writer/Penciller while displaying a comic details in the info page, but it is not implemented yet

Yeah it isn't implemented yet. I'm a bit conflicted about how to present that information in a way that isn't overly crowded, and I haven't landed on anything yet. Especially for people with libraries that have really rich metadata, it needs a good spot on the page

> I used to put some text into the <Volume> attribute, in order to fix an issue with Kavita that otherwise create me one volume per comic into a Series, and in Stump this field is restricted to Number, so it was failing processing the ComicInfo.xml

Ahhh okay, yeah gotcha. Stump follows the ComicInfo schema a big rigidly, and IIRC that defines `Volume` as an integer. If this fails in practice, i.e. there is a large amount of non-numeric volume metadata values out there, I can revisit this
